packages/clay-navigation-bar/src/hooks.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {removeCollapseHeight, setCollapseHeight} from './helpers';
+import {useEffect, useState} from 'react';
+
+export function useTransition(
+	visible: Boolean,
+	setVisible: any,
+	contentRef: React.RefObject<any>
+) {
+	const [transitioning, setTransitioning] = useState<Boolean>(false);
+
+	useEffect(() => {
+		if (transitioning) {
+			setCollapseHeight(contentRef);
+			if (visible) {
+				removeCollapseHeight(contentRef);
+			}
+		}
+	}, [transitioning]);
+
+	const handleTransitionEnd = (event: React.TransitionEvent) => {
+		if (event.target === contentRef.current && transitioning && !visible) {
+			setVisible(true);
+			setTransitioning(false);
+			removeCollapseHeight(contentRef);
+		} else if (event.target === contentRef.current) {
+			setVisible(false);
+			setTransitioning(false);
+		}
+	};
+
+	const willTriggerTransition = (event: React.MouseEvent | MouseEvent) => {
+		event.preventDefault();
+		if (visible && !transitioning) {
+			setCollapseHeight(contentRef);
+		}
+
+		if (!transitioning) {
+			setTransitioning(true);
+		}
+	};
+
+	return [transitioning, handleTransitionEnd, willTriggerTransition] as [
+		Boolean,
+		(event: React.TransitionEvent) => void,
+		(event: React.MouseEvent | MouseEvent) => void
```

Comment:
Is this necessary to cast all of these types? I would've thought that TS would be able to infer it from the type of  the variable.

Reply:
Removing these types Typescript isn't infering the type of these parameters. See the following screenshot:

<img width="564" alt="Screen Shot 2019-05-09 at 18 40 14" src="https://user-images.githubusercontent.com/7663513/57488601-fb142c00-7289-11e9-9d83-c3e7436db0ec.png">